### PR TITLE
Misc Iteration Cleanup

### PR DESCRIPTION
**Pull request by Gankro**
renames iterators, adds a cursor convenience.

Safety and curious API issues still abound; just slamming out some low-hanging fruit.

### CODE REVIEW

**Comment by Gankro** [comment]
Eh screw it, let's just lump in fixing up the iterators on everything else.

**Comment by Gankro** [comment]
Let no one say I'm *not* maintaining these crazy collections.

**Comment by Gankro** [comment]
rebased

r? @reem

File: src/tree/set.rs [hunk 1]
```diff
@@ -159,8 +159,8 @@ impl<T, C> TreeSet<T, C> where C: Compare<T> {
     /// ```
     #[inline]
     #[unstable = "matches collection reform specification, waiting for dust to settle"]
-    pub fn iter<'a>(&'a self) -> SetItems<'a, T> {
-        SetItems{iter: self.map.iter()}
+    pub fn iter<'a>(&'a self) -> Iter<'a, T> {
+        Iter{iter: self.map.iter()}
```
Comment (reem) [inline src/tree/set.rs]: Can you change the brace style on all these instantiations while you're at it?

File: src/tree/set.rs [hunk 2]
```diff
@@ -568,43 +568,43 @@ impl<T, C> TreeSet<T, C> where C: Compare<T> {
...
-pub struct SetItems<'a, T:'a> {
-    iter: Entries<'a, T, ()>
+pub struct Iter<'a, T:'a> {
+    iter: tree_map::Iter<'a, T, ()>
 }
 
 /// A lazy backward iterator over a set.
-pub struct RevSetItems<'a, T:'a> {
-    iter: RevEntries<'a, T, ()>
+pub struct RevIter<'a, T:'a> {
```
Comment (reem) [inline src/tree/set.rs]: Such boilerplate with all these wrapper types.

**Comment by reem** [comment]
You could call `seek_forward` and `seek_backward` `skip` and `skip_back` if you wanted to stay inline with existing iterators.

**Comment by Gankro** [comment]
I'm not sure if I want to treat cursors like iterators or like deques; the former are "biased" to forward, while the latter or "unbiased". So far I think I went with the latter strategy. 

For now I'm moderately happy to not give a fuck and come back to it with a more coherent vision once enough people complain :)

File: src/tree/set.rs (hunk 2 above)
Comment (Gankro) [inline src/tree/set.rs]: Yeah, it's not great. @aturon has some shadowy plans to make this specific usecase a bit more ergonomic (iterator newtypes).

**Comment by Gankro** [comment]
Fixed set braces

**Comment by reem** [comment]
Ok, sounds good.

I'm growing concerned with the "new type everything" approach, since it generates *a lot* of boilerplate. After a bit of use, do you think it's actually gaining us anything?

**Comment by Gankro** [comment]
It's strictly a future-proofing matter. If for some reason a shared iterator wants to be split, that's a breaking change. If it's all new-typed it's fine.

In the future I expect to have impl forwarding syntax, making it much more ergonomic. Something like:

```
struct Foo(Bar);

forward Foo {
  impl Iterator<u32>,
  impl Eq,
  frobnicate,
}
```
But probably won't get something like that for a while.

**Comment by Gankro** [comment]
r=reem

**Pull request merged**